scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala

```diff
@@ -1,62 +1,70 @@
 package com.twitter.scalding.beam_backend
 
+import cascading.flow.FlowDef
 import com.stripe.dagon.Rule
 import com.twitter.scalding.Execution.{ToWrite, Writer}
 import com.twitter.scalding.typed._
-import com.twitter.scalding.{CFuture, CancellationHandler, Config, Execution, ExecutionCounters}
-import java.io.InputStream
+import com.twitter.scalding.{CFuture, CancellationHandler, Config, Execution, ExecutionCounters, Mode}
 import java.nio.channels.Channels
-import java.util.concurrent.atomic.AtomicLong
 import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicLong
 import org.apache.beam.sdk.Pipeline
 import org.apache.beam.sdk.coders.Coder
 import org.apache.beam.sdk.io.FileSystems
 import scala.annotation.tailrec
 import scala.collection.convert.decorateAsScala._
 import scala.concurrent.{ExecutionContext, Future}
+import scala.collection.JavaConversions._
+
+case class TempSource[A](path: String, coder: Coder[A]) extends TypedSource[A] {
+  def error = sys.error("beam sources don't work in cascading")
+  def converter[U >: A] = error
+  def read(implicit flowDef: FlowDef, mode: Mode) = error
+}
 
 class BeamWriter(val beamMode: BeamMode) extends Writer {
   private val state = new AtomicLong()
 
   private val sourceCounter: AtomicLong = new AtomicLong(0L)
-  private val iterablePipes: scala.collection.concurrent.Map[TypedPipe[_], (Coder[_], String)] =
-    new ConcurrentHashMap[TypedPipe[_], (Coder[_], String)]().asScala
+
+  val tempSources: scala.collection.concurrent.Map[TypedPipe[_], TempSource[_]] =
+    new ConcurrentHashMap[TypedPipe[_], TempSource[_]]().asScala
 
   override def start(): Unit = ()
 
-  override def finished(): Unit = ()
+  override def finished(): Unit = {
+    // `FileSystems.delete` fails to delete dir as it contains files, hence we delete the files in the dir
+    // There is a temp subdir with name starting with "." which is not matched by `match`.
+    // So currently a single empty dir is left behind.
+    val resources = tempSources.values
+      .map(ts => s"${ts.path}*")
+      .flatMap(path => FileSystems.`match`(path).metadata().map(_.resourceId()))
+    FileSystems.delete(resources.toList)
+  }
 
   def getForced[T](conf: Config, initial: TypedPipe[T])(implicit
       cec: ExecutionContext
-  ): Future[TypedPipe[T]] = ???
+  ): Future[TypedPipe[T]] =
+    tempSources.get(initial) match {
+      case Some(source) => Future.successful(TypedPipe.from(source).asInstanceOf[TypedPipe[T]])
```

Comment:
actually, can we add the other case here:

`case None => Future.failed(new IllegalStateException(s"TypedPipe = $inital, has not yet been forced")`